server/src/main/java/org/elasticsearch/cluster/routing/ShardRouting.java

```diff
@@ -39,6 +39,13 @@ public final class ShardRouting implements Writeable, ToXContentObject {
 
     private final ShardId shardId;
     private final String currentNodeId;
+    /**
+     * This field contains
+     * - node id this shard is relocating to iff state == RELOCATING
+     * - node id this shard is relocating from iff state == INITIALIZING/UNASSIGNED and this is relocation target
```

Comment:
I don't think we can have a `relocatingNodeId` on an `UNASSIGNED` shard. We should be able to assert that.